cedar-policy-validator/src/schema.rs

```diff
@@ -2571,23 +2767,1487 @@ mod test {
+    /// add a `MyType` use of kind C2X1 (for any X), returning the new schema.
+    fn C2X1_json(mut schema: serde_json::Value) -> serde_json::Value {
+        schema["NS1"]["commonTypes"]["E"] = json!({
+            "type": "Record",
+            "attributes": {
+                "foo": { "type": "Entity", "name": "NS1::MyType" }
+                }
+        });
+        schema
+    }
+
+    /// Given a starting JSON schema (e.g., from `a1_json()`),
+    /// add a `MyType` use of kind C2X2 (for any X), returning the new schema.
+    fn C2X2_json(mut schema: serde_json::Value) -> serde_json::Value {
+        schema["NS1"]["commonTypes"]["E"] = json!({
+            "type": "Record",
+            "attributes": {
+                "foo": { "type": "NS1::MyType" }
+            }
+        });
+        schema
+    }
+
+    /// Generate human-schema syntax for a `MyType` use of kind C3.
+    fn C3_human() -> &'static str {
+        r#"type E = { foo: NS2::MyType };"#
+    }
+
+    /// Given a starting JSON schema (e.g., from `a1_json()`),
+    /// add a `MyType` use of kind C3X1 (for any X), returning the new schema.
+    fn C3X1_json(mut schema: serde_json::Value) -> serde_json::Value {
+        schema["NS1"]["commonTypes"]["E"] = json!({
+            "type": "Record",
+            "attributes": {
+                "foo": { "type": "Entity", "name": "NS2::MyType" }
+                }
+        });
+        schema
+    }
+
+    /// Given a starting JSON schema (e.g., from `a1_json()`),
+    /// add a `MyType` use of kind C3X2 (for any X), returning the new schema.
+    fn C3X2_json(mut schema: serde_json::Value) -> serde_json::Value {
+        schema["NS1"]["commonTypes"]["E"] = json!({
+            "type": "Record",
+            "attributes": {
+                "foo": { "type": "NS2::MyType" }
+            }
+        });
+        schema
+    }
+
+    // ####
+    //
+    // For explanations of all of the below tests and their naming, see comments
+    // on this module.
+    //
+    // ####
+
+    // human versions
+    #[test]
+    fn A1a1_human() {
+        assert_parses_successfully_human(&a1_human(A1_human()));
+    }
+    #[test]
+    fn A1a2_human() {
+        assert_parses_successfully_human(&a2_human(A1_human()));
+    }
+    #[test]
+    fn A1b1_human() {
+        assert_parses_successfully_human(&b1_human(A1_human()));
+    }
+    #[test]
+    fn A1b2_human() {
+        assert_parses_successfully_human(&b2_human(A1_human()));
+    }
+    #[test]
+    fn A1c_human() {
+        assert_parse_error_human(
+            &c_human(A1_human()),
+            &ExpectedErrorMessageBuilder::error("failed to resolve type: MyType")
+                .help(
+                    "neither `NS1::MyType` nor `MyType` refers to anything that has been declared",
+                )
+                // TODO: why don't we have an underline here? (Uncommenting this produces test failure)
+                //.exactly_one_underline("MyType")
```

Comment:
Investigated this briefly and I think it's just #174 again -- the JSON structures don't carry source locations, and since the human processing works by first converting to the JSON structures, we lose the source locations at that point.  Note this is a regression, as previously some errors were detected during typename resolution before converting to JSON, and came with source locations, whereas now, typename resolution happens after converting to JSON and thus loses source locations.

Reply:
Updated these TODO comments with an explanation and pointing to #174.  In my opinion we need to accept this regression; merging this PR and fixing #579 is more important.

Reply:
I think that sounds fine. Although, this could be tracked separately from #174. The work described there covers capturing source information out of serde, but we only need to carry source info in the schema  (without capturing it tn the JSON case) to fix this regression.

Reply:
Created #1084 for this.

Reply:
Updated the comments to point to 1084 instead of 174.  Resolving this.